core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterSuite.scala

```diff
@@ -184,4 +184,20 @@ class DiskBlockObjectWriterSuite extends SparkFunSuite with BeforeAndAfterEach {
     writer.close()
     assert(segment.length === 0)
   }
+
+  test("calling closeAndDelete() on a partial write file") {
+    val (writer, file, writeMetrics) = createWriter()
+
+    writer.write(Long.box(20), Long.box(30))
+    val firstSegment = writer.commitAndGet()
+    assert(firstSegment.length === file.length())
+    assert(writeMetrics.bytesWritten === file.length())
+
+    writer.write(Long.box(40), Long.box(50))
+
+    writer.closeAndDelete()
+    assert(!file.exists())
+    assert(writeMetrics.bytesWritten === firstSegment.length)
```

Comment:
I add a new case, the current value of `writeMetrics` is the value after the last `commitAndGet` now. I think we should reset values of `writeMetrics` to 0 because the file no longer exists. What do you think? @attilapiros

Reply:
A little hesitant

Reply:
In the test it must be 0 but we should decrease the metric by the sum of all records written by the writer.

What about counting how many records were committed successfully and decrement the `recordsWritten` by the sum of committed records and current `numRecordsWritten`?

Reply:
Like 9db7115 changed?

Reply:
This seems to increase the memory consumption of a `Long` variable, or do we add a `reset` api for `ShuffleWriteMetricsReporter`?

Reply:
> Like [9db7115](https://github.com/apache/spark/commit/9db7115fc980c80ee517f46e6844b39d76c93559) changed?

Not exactly. I would keep track of the commited records and not the total written records.
When the committed records is counted then you do not need to increase the new var every time when a record is written but only when a huge number of records are committed. 
So this line is not needed:
https://github.com/apache/spark/blob/9db7115fc980c80ee517f46e6844b39d76c93559/core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala#L329
But you need to increase the new var after the line 232 before the reset of `numRecordsWritten `: https://github.com/apache/spark/blob/9db7115fc980c80ee517f46e6844b39d76c93559/core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala#L231-L233

This way when the file is remove you can decrease the metric with the sum of the new var and `numRecordsWritten` at:
https://github.com/apache/spark/blob/9db7115fc980c80ee517f46e6844b39d76c93559/core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala#L291

Reply:
> This seems to increase the memory consumption of a `Long` variable, or do we add a `reset` api for `ShuffleWriteMetricsReporter`?

I might miss something. Why do you think an extra `Long` matters here regarding memory consumption?

Reply:
a1bad58 refactored the code